libsolidity/lsp/Transport.h

```diff
@@ -0,0 +1,100 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+#pragma once
+
+#include <json/value.h>
+
+#include <functional>
+#include <iosfwd>
+#include <map>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <variant>
+
+namespace solidity::lsp
+{
+
+using MessageID = Json::Value;
+
+enum class ErrorCode
+{
+	// Defined by JSON RPC
+	ParseError = -32700,
+	MethodNotFound = -32601,
+	InvalidParams = -32602,
+	InternalError = -32603,
+
+	// Defined by the protocol.
+	RequestFailed = -32803
+};
+
+/**
+ * Transport layer API
+ *
+ * The transport layer API is abstracted so it users become more testable as well as
+ * this way it could be possible to support other transports (HTTP for example) easily.
+ */
+class Transport
```

Comment:
I think that naming there could be clearer. Initially these names made me think that `Transport` is some generic transport implementation and `JSONTransport` builds on top of it but it looks like the former is actually just an interface and the latter is a concrete implementation based on iostreams. Also, the naming suggests that `JSONTransport` operates on JSON messages while `Transport` is generic but it's not the case - both deal with JSON messages.

Maybe we could name them like this?
- `Transport` -> `JSONTransport`
- `JSONTransport` -> `IOStreamJSONTransport`

```suggestion
 * Interface describing a transport layer for exchanging JSON messages.
 *
 * The transport layer API is abstracted so it users become more testable as well as
 * this way it could be possible to support other transports (HTTP for example) easily.
 */
class JSONTransport
```